exonum/src/runtime/mailbox/blockchain_secretary.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module contains the entities capable of processing messages received through
+//! `BlockchainMailbox`.
+
+// TODO only for development purposes, remove it later
+use futures::Future;
+
+use exonum_merkledb::Fork;
+
+use crate::runtime::{
+    dispatcher::{Dispatcher, Error as DispatcherError},
+    error::{catch_panic, ExecutionError},
+    CallInfo, InstanceSpec,
+};
+
+use super::{Action, AfterRequestCompleted, BlockchainMailbox, Notification};
+
+/// Mailbox processing context.
+///
+/// This enum defines the rules of requests processing:
+/// if `MailboxContext` is the `TxExecution`, then any failure appeared during
+/// processing will be reported (so blockchain core will be able to revert the transaction).
+/// Otherwise any failure will be simply ignored.
+#[derive(Debug, Clone)]
+pub enum MailboxContext {
+    /// Mailbox processing happening during the transaction execution process.
+    TxExecution(CallInfo),
+    /// Mailbox processing outside of the transaction execution process.
+    OutsideTxExecution,
+}
+
+/// `BlockchainSecretary` is an intermediate entity capable of processing the requests
+/// from services to `Blockchain`, and delegating the allowed actions to the dispatcher.
+/// Also `BlockchainSecretary` creates notifications for `Blockchain` about performed actions
+/// so `Blockchain` will be able to do any follow-up things if they are needed.
+#[derive(Debug)]
+pub struct BlockchainSecretary {
+    context: MailboxContext,
+}
+
+impl BlockchainSecretary {
+    /// Creates a new `BlockchainSecretary` instance.
+    pub fn new(context: MailboxContext) -> Self {
+        Self { context }
+    }
+
+    /// Processes requests within immutable context.
+    /// Since any failures won't affect the blockchain state, they are simply ignored.
+    pub fn process_requests(&self, mailbox: &mut BlockchainMailbox, dispatcher: &mut Dispatcher) {
+        // During requests processing new requests may appear.
+        loop {
```

Comment:
Well, if we will talk about hypothetical situations, I can imagine a scenario when updating config of service A depends on updating of config of service B which depend on updating of config of service A.

But in that scenario such a loop will appear anyway, directly or indirectly (e.g. through recursion), and I don't think that it can be avoided from the side of blockchain.

Talking about correct scenarios, I can not imagine it.

However, I can add an iteration counter (e.g. 4) on which loop will be broken. Again, it's unlikely to be ever achieved, but we can do it for the glory of Defencive Programming.

Will it be ok for you?